### DISCUSSION TITLE
mmm.fit Method Hangs Indefinitely During Model Fitting

### DISCUSSION THREAD

**Discussion by stdp95** [pymc-labs/pymc-marketing#1255] [General]
I’m encountering an issue where the `mmm.fit` method seems to run indefinitely and does not complete. Here’s the call I’m using:

```python
mmm.fit(X_train, y_train, target_accept=0.95, chains=4, draws=2000, tune=1000)
```
### Problem Details

-  The process appears to hang during the fitting phase, with no progress or errors reported.
- I’ve ensured that my input data (X_train and y_train) are properly preprocessed and scaled as required.

### Question
Has anyone else encountered this behavior? If so, are there any debugging steps or fixes I could try? Any insights into what might be causing the infinite runtime would be greatly appreciated.

PS: My config for reference
```python
mmm_config = {
    'intercept': {'dist': 'Normal', 
                  'kwargs': {'mu': 0, 'sigma': 2}},
   
    'beta_channel': {'dist': 'LogNormal',
                     'kwargs': {'mu': 1, 'sigma': prior_sigma}},  
    
    'alpha': {'dist': 'Beta',
              'kwargs': {'alpha': 1, 'beta': 3}},
    
    'lam': {'dist': 'Gamma',
            'kwargs': {'alpha': 3, 'beta': 1}},
    
    'likelihood': {'dist': 'Normal',
                   'kwargs': {'sigma': {'dist': 'HalfNormal', 'kwargs': {'sigma': 2}}}},
    
    'gamma_control': {'dist': 'Normal',
                      'kwargs': {'mu': 0, 'sigma': 1}},
    
    'gamma_fourier': {'dist': 'Laplace',
                      'kwargs': {'mu': 0, 'b': 1}}
}
```

**Comment by stdp95** [comment]
When I previously tried the following configuration with fewer draws and a shorter tuning phase:

```python
mmm.fit(X_train, y_train, target_accept=0.95, chains=4, draws=1000, tune=500)
```
I received the following warning message:

```ruby
WARNING:pymc.stats.convergence:Chain 0 reached the maximum tree depth. 
Increase `max_treedepth`, increase `target_accept` or reparameterize.
```
Do you think this warning might be related to the issue I'm facing with the indefinite runtime? Could this indicate a problem with the priors or model configuration that’s causing sampling difficulties?

**Reply by AlfredoJF** [reply]
@stdp95 Have you tried [Other NUTS samplers](https://github.com/pymc-labs/pymc-marketing/blob/main/docs/source/notebooks/general/other_nuts_samplers.ipynb)? PyMC's default sampler is quite slow IMO. Numpyro and Nutpie have these nice progress bars for each chain and estimations of runtime.

**Reply by stdp95** [reply]
I hadn’t tried other NUTS samplers before, but after your suggestion, I gave NumPyro a try, and it seems to work much better. Thanks for the recommendation.

**Reply by williambdean** [reply]
Thanks @AlfredoJF for the suggestion. nutpie is also a good alternative

**Reply by stdp95** [reply]
Since NumPyro has worked well for my needs, I haven’t prioritized exploring nutpie. While it seems like a good alternative, its reliance on Rust dependencies might complicate the setup in Databricks.

**Comment by williambdean** [comment]
Can you share the pymc marketing, pymc version as well as how you installed and your machine

**Reply by stdp95** [reply]
I installed pymc-marketing using pip, which also installed pymc as a dependency. Here are the versions I am using:

- pymc-marketing version: 0.8.0
- pymc version: 5.15.1

I am running this on Databricks.

**Reply by williambdean** [reply]
Are you able to sample simple models with `pymc`? Does that hang? 

```python
import numpy as np
import pymc as pm

seed = sum(map(ord, "Simple model"))
rng = np.random.default_rng(seed)

mu = np.array([1, 2, 3])
sigma = 0.25

n = 500
gdp = pm.Normal.dist(mu=mu, sigma=sigma, shape=(n, 3))
data = pm.draw(gdp, random_seed=rng)

coords = {"channel": ["A", "B", "C"], "idx": range(n)}
with pm.Model(coords=coords) as model:
    mu = pm.Normal("mu", dims="channel")
    sigma = pm.HalfNormal("sigma")
    pm.Normal("obs", mu=mu, sigma=sigma, observed=data, dims=("idx", "channel"))
   
idata = pm.sample(model=model)
```

Do you get any warnings or messages at any point?

**Reply by stdp95** [reply]
Nope, it ran fine without any issues. I didn’t encounter any warnings or messages during the sampling process.

**Comment by williambdean** [comment]
How does it do with the default priors? Are you changing the MMM configurations in any other way?

**Reply by stdp95** [reply]
Even with the default priors, the process is taking an unusually long time. I ended up canceling the run after 3 hours because it stalled at 28%, which was the same progress it had made in the first 10 minutes. I’m pretty sure that I haven’t made any changes to the MMM configurations.

**Comment by williambdean** [comment]
Can you show us the graphviz output?

**Reply by stdp95** [reply]
I’m not sure if it’s possible to generate the Graphviz output before running the fit process. Typically, I’ve been able to generate it only after fitting the model. Since I’m currently stuck during the fit process, I haven’t been able to produce the Graphviz output in this instance.

If there’s a way to generate the output prior to fitting, I’d appreciate any guidance on how to proceed!

**Reply by williambdean** [reply]
You can use the build_model with X and y then use graphviz(?) method. 

You can also try the sample_prior_predictive method to see what your model generates

**Reply by stdp95** [reply]
[model_graph (1).pdf](https://github.com/user-attachments/files/18025999/model_graph.1.pdf)

**Reply by williambdean** [reply]
Thanks for sharing. Not that the "beta_channel", "alpha", and "lam" are not in the graph. Those are not the names of the parameter in the model. Unfortunately, it is a silent when keys are not used

**Reply by stdp95** [reply]
Thank you for pointing that out! I hadn’t realized that the keys beta_channel, alpha, and lam are not directly recognized as parameters in the model. Since the model remains silent when unused keys are present, it would be really helpful if you could share what the correct parameter names should be. This will ensure I align my configuration properly. Appreciate your guidance!

**Reply by stdp95** [reply]
I assume these are the correct parameter names:

```python
{
    'intercept': Prior("Normal", mu=0, sigma=2),
    'likelihood': Prior("Normal", sigma=Prior("HalfNormal", sigma=2)),
    'gamma_control': Prior("Normal", mu=0, sigma=2, dims="control"),
    'gamma_fourier': Prior("Laplace", mu=0, b=1, dims="fourier_mode"),
    'adstock_alpha': Prior("Beta", alpha=1, beta=3, dims="channel"),
    'saturation_lam': Prior("Gamma", alpha=3, beta=1, dims="channel"),
    'saturation_beta': Prior("HalfNormal", sigma=2, dims="channel")
}

**Reply by williambdean** [reply]
Yes, appears right. I created an issue for this

**Reply by stdp95** [reply]
I updated the priors, but the progress bar remains stuck at 28% with no further progress.
[model graph.pdf](https://github.com/user-attachments/files/18030478/model.graph.pdf)

**Reply by williambdean** [reply]
But no with the numpyro sampler though? We must not use the default C backend. 

How does that sampling do? How are the parameter estimates

**Comment by williambdean** [comment]
Can you provide some information about your data, number of rows, numer of channels, number of controls, etc

**Reply by stdp95** [reply]
The dataset contains weekly data spanning 105 weeks. It includes 8 marketing channels and 1 control variable.
